File: Algorithms/200. Number of Islands/solution.js
var numIslands = function (grid) {
  if (!grid || grid.length === 0) return 0;

  let count = 0;
  const rows = grid.length;
  const cols = grid[0].length;

  function dfs(i, j) {
    if (i < 0 || j < 0 || i >= rows || j >= cols || grid[i][j] !== "1") return;

    grid[i][j] = "0";

    dfs(i + 1, j);
    dfs(i - 1, j);
    dfs(i, j + 1);
    dfs(i, j - 1);
  }

  for (let i = 0; i < rows; i++) {
    for (let j = 0; j < cols; j++) {
      if (grid[i][j] === "1") {
        count++;
        dfs(i, j);
      }
    }
  }

  return count;
};

File: Algorithms/200. Number of Islands/solution.cpp
class Solution
{
public:
    int numIslands(vector<vector<char>> &grid)
    {
        if (grid.empty())
            return 0;

        int rows = grid.size();
        int cols = grid[0].size();
        int ans = 0;

        vector<pair<int, int>> dirs = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};
        queue<pair<int, int>> q;

        for (int i = 0; i < rows; i++)
        {
            for (int j = 0; j < cols; j++)
            {
                if (grid[i][j] == '1')
                {
                    ans++;
                    q.push({i, j});
                    grid[i][j] = '0';

                    while (!q.empty())
                    {
                        auto [x, y] = q.front();
                        q.pop();

                        for (auto [dx, dy] : dirs)
                        {
                            int nx = x + dx;
                            int ny = y + dy;

                            if (nx >= 0 && ny >= 0 && nx < rows && ny < cols &&
                                grid[nx][ny] == '1')
                            {
                                q.push({nx, ny});
                                grid[nx][ny] = '0';
                            }
                        }
                    }
                }
            }
        }
        return ans;
    }
};
